### DISCUSSION TITLE
How can I authenticate to the Azure CR?

### DISCUSSION THREAD

**Discussion by konstantinblaesi** [renovatebot/renovate#10180] [Request Help]
All our images live within our teams registry, e.g. team.azurecr.io. I have created a token for our registry giving me a username and password. I tested them, `docker login -u $USER -p $PASS team.azurecr.io` works. How can I make renovate use these credentials to check whether our base images need updating?

**Comment by rarkins** [comment]
Configure a `hostRules` entry with `hostType` (`docker`), `matchHost`, `username` and `password`. Docs: https://docs.renovatebot.com/private-modules/

**Reply by konstantinblaesi** [reply]
Thanks, that already helped. I currently face some weird behavior of renovate + azure that I don't understand yet. I am currently debugging renovate locally, maybe you have an idea what's going on here.

`curl -v https://some-team.azurecr.io/v2/` answers with this response header
> www-authenticate: Bearer realm="https://some-team.azurecr.io/oauth2/token",service="some-team.azurecr.io"

which to me sounds like I would want [this renovate code path to be taken](https://github.com/renovatebot/renovate/blob/main/lib/datasource/docker/common.ts#L95)

Instead while debugging I see a different response header returned by azure
> www-authenticate: Basic realm="Azure Container Registry"

I have verified that I can get an 'access_token' using curl like this `curl -v -u "some-user:some-pass" https://some-team.azurecr.io/oauth2/token?service=some-team.azurecr.io&scope=repository:some-image:pull`. The response is as expected something like this
>{"access_token":"some-token"}

Because the www-authenticate reponse header value renovate tries basic authentication and fails instead of requesting an oauth2 token. Renovate logs this among other things in debug mode:
> DEBUG: Using Basic auth for docker registry some-image (repository=some-user/some-repo)
> DEBUG: Unauthorized docker lookup (repository=some-user/some-repo)
>        "registryHost": "https://some-team.azurecr.io",
>        "dockerRepository": "some-image"
> DEBUG: Response code 401 (Unauthorized) (repository=some-user/some-repo)

**Comment by viceice** [comment]
Try this host rule:
```json
{
  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "some-team.azurecr.io",
      "username": "some-user",
      "password": "some-pass"
    }
  ]
}
```

**Reply by konstantinblaesi** [reply]
Yes I am already using exactly this host rule and I think I have found the root cause for this response header behavior. I can replicate it with curl:

`curl -v -u "some-user:some-pass" https://some-team.azurecr.io/v2/` results in response header

>www-authenticate: Basic realm="Azure Container Registry"

while `curl -v https://some-team.azurecr.io/v2/` results in response header
>www-authenticate: Bearer realm="https://some-team.azurecr.io/oauth2/token",service="some-team.azurecr.io"

As a quick hack I just tried this and now it hits the 'oauth2 token' code path as expected (access_token received and used):
>lib/util/http/index.ts
![image](https://user-images.githubusercontent.com/5871169/119887919-0734ba00-bf35-11eb-9920-838432f020fd.png)

renovate debug logs changed a bit:
```
DEBUG: Datasource unauthorized (repository=some-user/some-repo)
       "datasource": "docker",
       "lookupName": "some-team.azurecr.io/some/image",
       "url": "https://some-team.azurecr.io/v2/some/image/tags/list?n=10000"
DEBUG: Failed to look up dependency some-team.azurecr.io/some/image (repository=some-user/some-repo, packageFile=Dockerfile, dependency=some-team.azurecr.io/some/image)`
```
ACR denies the GET request to https://some-team.azurecr.io/v2/some/image/tags/list?n=10000 with my bearer token, I need to look into this further. Maybe a configuration issue in azure I am not sure. The ACR response is
```
{
    "errors": [
        {
            "code": "UNAUTHORIZED",
            "message": "authentication required, visit https://aka.ms/acr/authorization for more information.",
            "detail": [
                {
                    "Type": "repository",
                    "Name": "some/image",
                    "Action": "metadata_read"
                }
            ]
        }
    ]
}
```
I tried with metadata/read or all permissions for the scopes in azure, no change.

Am I using ACR wrong with renovate or is this a bug ? If it is, should renovate make sure[ this URL is requested](https://github.com/renovatebot/renovate/blob/main/lib/datasource/docker/common.ts#L59) without the username/password basic auth credentials from my host rule as I've done with my small code change ?

**Reply by rarkins** [reply]
Renovate's approach to authentication is pretty simple. e.g. if there's a `token` in `hostRules` then it gets mapped to a `Bearer` authorization, etc. The best thing you could do is keep experimenting with `curl` and work out the sequence that works so that we can determine if we need to update any of our logic.

**Reply by Lingkar** [reply]
@rarkins I have done some curl experimenting and the following sequence works for me:
Does that help with understanding compatibility with the current logic?

The scope found is: repository:some/image:metadata_read

```
tenant={{TENANT_ID}}
registry={{ACR_HOST}}
client_id={{SERVICE_PRINCIPAL_ID}}
client_secret={{SERVICE_PRINCIPAL_SECRET}}
scope="https://management.azure.com//.default"
path="/v2/some/image/tags/list?n=1000"

aad_access_token=$(
    curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
        -d "client_id=$client_id" \
        -d "grant_type=client_credentials" \
        -d "scope=$scope" \
        -d "client_secret=$client_secret" https://login.microsoftonline.com/$tenant/oauth2/v2.0/token | jq -r .access_token
)

echo "AAD access token:"
echo $aad_access_token
echo

acr_refresh_token=$(
    curl -s -X POST -H "Content-Type: application/x-www-form-urlencoded" \
        -d "grant_type=access_token&service=$registry&tenant=$tenant&access_token=$aad_access_token" \
        https://$registry/oauth2/exchange | jq -r .refresh_token
)

echo "ACR refresh token:"
echo $acr_refresh_token
echo

scope=$(curl -s https://$registry$path | jq -r '.errors[].detail[] | "\(.Type):\(.Name):\(.Action)"')

echo "Scope $scope"
echo

acr_access_token=$(
    curl -s -X POST -H "Content-Type: application/x-www-form-urlencoded" \
        -d "grant_type=refresh_token&service=$registry&scope=$scope&refresh_token=$acr_refresh_token" \
        https://$registry/oauth2/token | jq -r .access_token
)

echo "ACR access token:"
echo $acr_access_token
echo

curl -H "Authorization: Bearer $acr_access_token" https://$registry$path
```

Usefull docs:
https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#calling-post-oauth2exchange-to-get-an-acr-refresh-token
